auth/auth.go

```diff
@@ -0,0 +1,131 @@
+package auth
+
+import (
+	"crypto/ed25519"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/brave/go-sync/datastore"
+	"github.com/brave/go-sync/utils"
+)
+
+var (
+	timestampMaxDuration int64 = 120 * 1e3 // Milliseconds, modifiable in tests.
+)
+
+const (
+	tokenMaxDuration int64  = 86400 * 1e3 // Milliseconds
+	bearerPrefix     string = "Bearer "
+	nRandBytes       int    = 32 // Number of random bytes used to generate the access token.
+)
+
+// Request is a struct used for authenication requests.
+type Request struct {
+	PublicKey       string `json:"public_key"`
+	Timestamp       string `json:"timestamp"`
+	SignedTimestamp string `json:"signed_timestamp"`
+}
+
+// Response is a struct used for authenication responses.
+type Response struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int64  `json:"expires_in"`
+}
+
+// generateToken generates n random bytes and encoded it as base64 string.
+func generateToken(n int) (string, error) {
+	bytes := make([]byte, n)
+	_, err := rand.Read(bytes)
+	if err != nil {
+		return "", err
+	}
+
+	return base64.URLEncoding.EncodeToString(bytes), nil
+}
+
+// Authenticate validates client's auth requests and provides the reply.
+func Authenticate(r *http.Request, db datastore.Datastore) (string, []byte, error) {
+	var rsp []byte
+
+	err := r.ParseForm()
+	if err != nil {
+		return "", nil, err
+	}
+	req := &Request{
+		PublicKey:       r.PostFormValue("client_id"),
+		Timestamp:       r.PostFormValue("timestamp"),
+		SignedTimestamp: r.PostFormValue("client_secret"),
+	}
+
+	// Verify the signature.
+	publicKey, err := hex.DecodeString(req.PublicKey)
+	if err != nil {
+		return "", nil, err
+	}
+	timestampBytes, err := hex.DecodeString(req.Timestamp)
+	if err != nil {
+		return "", nil, err
+	}
+	signedTimestamp, err := hex.DecodeString(req.SignedTimestamp)
+	if err != nil {
+		return "", nil, err
+	}
+	if !ed25519.Verify(publicKey, timestampBytes, signedTimestamp) {
+		return "", nil, fmt.Errorf("signature verification failed")
+	}
+
+	var timestamp int64
+	timestamp, err = strconv.ParseInt(string(timestampBytes), 10, 64)
+	if err != nil {
+		return "", nil, fmt.Errorf("parse timestamp error")
+	}
+
+	// Verify the timestamp is not outdated
+	if utils.UnixMilli(time.Now())-timestamp > timestampMaxDuration {
```

Comment:
Not clear on the purpose of this check... It seems that this service will accept any timestamp (signed by the user) so, why can't I just make my timestamp 2150-01-01... and sign it, and hand it to authorization (based on the code the user will get back a token if they sign a timestamp in the far future).  I am unclear on the purpose of the /timestamp endpoint...

It seems the point of authenticate is to make sure that this user-agent has knowledge of the private key, and is able to sign anything.  I don't feel like this needs to be an interactive protocol.

Reply:
Reading the spec it seems this interactive protocol is here to validate that an intercepted old signature is unlikely to be re-used... i still think you need to check the absolute value here.  Might even be worth in timestamp handing a server signed timestamp back, which can be verified that they got the timestamp from the server and didnt just make it up

Reply:
> Reading the spec it seems this interactive protocol is here to validate that an intercepted old signature is unlikely to be re-used

that's correct. although we assume TLS makes a MITM attack infeasible, we want to ensure that the signature isn't a permanently-valid token in case it ever gets compromised.

Reply:
another way of doing it is with a nonce, but that seemed harder to me than doing it with a timestamp because then the server has to keep state about what nonces have been used before

Reply:
I don't think we'll check the exact timestamp here, the main reason we provide the timestamp from server is to avoid client side clock synchronization issues as mentioned in the spec. I think the main thing we enforce here is whether a client has a valid key pair to sign the message. So if the client could sign it, even the time is in the future, it is still consider passed in our protocol. And the reason of checking if it is exceeding the max duration is what @diracdeltas said above, we want to prevent someone replays a message they intercepted before and tries to get the token.

Reply:
@diracdeltas thanks you for the background.  Would it be prudent to take the absolute value of the timestamp subtraction before comparing it to the max life duration of the timestamp (for cases where the timestamp for some reason is in the future):

```go
if math.Abs(utils.UnixMilli(time.Now())-timestamp)) > timestampMaxDuration
```
Optionally, we could change the "timestamp" endpoint to return the timestamp and a signature created by the server, of the timestamp.  On Authenticate we would require the timestamp, the server signature and the user signature.  

That way we would be able to prove that the user knows their private key, and the user signed a timestamp that we can prove we gave them. 

I might be overthinking this though :)

Reply:
I think it's a good point to use `math.Abs` here just to be extra safe. Not sure about if we would like to change the authenticate protocol by adding server signature though, it will need some more client side work on this too if we do want it. cc @darkdh

Reply:
i don't think doing the absolute value check would hurt. if i understand correctly, correctly-functioning computer clocks can drift by seconds per day in either direction so that can add up if they're not syncing with NTP servers for some reason. (but at some point, other things like TLS certificate validation will break if the clock is too out of sync - so presumably someone would notice and fix it)

i think signing a timestamp by the server (versus providing an unsigned server timestamp) is only useful if we're trying to prevent clients who might lie about the time? i'm not sure i see a case where clients would do that. if an attacker can get a signed client timestamp to replay, they can presumably also get the signed server timestamp using whatever method they used to get the signed client timestamp.

Reply:
Resolve this here because this step is now removed.